### PR TITLE
AC-346: Implement RxJava in LocationDAO

### PR DESCRIPTION
**Pull request by shivtej1505**

### CODE REVIEW

File: openmrs-client/src/main/java/org/openmrs/mobile/activities/login/LoginPresenter.java [hunk 1]
```diff
@@ -212,15 +213,18 @@ public void onFailure(Call<Results<Location>> call, Throwable t) {
...
-            List<Location> locations = LocationDAO.getLocations();
-            if (locations.size() > 0) {
-                loginView.initLoginForm(locations, url);
-                loginView.setLocationErrorOccurred(false);
-            } else {
-                loginView.showToast("Network not available.", ToastUtil.ToastType.ERROR);
-                loginView.setLocationErrorOccurred(true);
-            }
-            loginView.hideLoadingAnimation();
+            LocationDAO.getLocations()
```
Comment (AdamGrzybkowski) [inline openmrs-client/src/main/java/org/openmrs/mobile/activities/login/LoginPresenter.java]: missing addSubscription()

File: openmrs-client/src/main/java/org/openmrs/mobile/api/retrofit/VisitApi.java [hunk 2]
```diff
@@ -157,7 +157,13 @@ public void startVisit(final Patient patient, @Nullable final StartVisitResponse
...
         visit.setStartDatetime(DateUtils.convertTime(System.currentTimeMillis(), DateUtils.OPEN_MRS_REQUEST_FORMAT));
         visit.setPatient(patient);
-        visit.setLocation(LocationDAO.findLocationByName(OpenMRS.getInstance().getLocation()));
+
+        LocationDAO.findLocationByName(OpenMRS.getInstance().getLocation())
+                .observeOn(AndroidSchedulers.mainThread())
+                .subscribe(location -> {
+                    visit.setLocation(location);
+                });
+
```
Comment (AdamGrzybkowski) [inline openmrs-client/src/main/java/org/openmrs/mobile/api/retrofit/VisitApi.java]: this way visit.setLocation() might be called after doing retrofit call. I see there is a lot of usage of  this method where we don't want async task so i would leave it as it was.

File: openmrs-client/src/main/java/org/openmrs/mobile/dao/VisitDAO.java [hunk 3]
```diff
@@ -191,21 +191,28 @@ private boolean updateVisit(Visit visit, long visitID, long patientID) {
...
                         int visitStop_CI = cursor.getColumnIndex(VisitTable.Column.STOP_DATE);
                         int visitPatientID_CI = cursor.getColumnIndex(VisitTable.Column.PATIENT_KEY_ID);
-                        visit = new Visit();
+                        Visit visit = new Visit();
                         visit.setUuid(cursor.getString(visitUUID_CI));
                         visit.setId(cursor.getLong(visitID_CI));
                         visit.setVisitType(new VisitType(cursor.getString(visitType_CI)));
-                        visit.setLocation(LocationDAO.findLocationByName(cursor.getString(visitPlace_CI)));
+
+                        LocationDAO.findLocationByName(cursor.getString(visitPlace_CI))
```
Comment (AdamGrzybkowski) [inline openmrs-client/src/main/java/org/openmrs/mobile/dao/VisitDAO.java]: this is bad for sure... visit.setLocation() cannot be called in async here. Same for all methods below

File: openmrs-client/src/test/java/org/openmrs/mobile/test/presenters/PatientDashboardVisitsPresenterTest.java [hunk 4]
```diff
@@ -157,8 +157,10 @@ public void shouldStartVisit_failure(){
 
 
     private void createMocksForStartVisit() {
+        Observable observable = Observable.create((Observable.OnSubscribe<Location>) subscriber -> {});
         PowerMockito.when(OpenMRS.getInstance()).thenReturn(openMRS);
-        PowerMockito.when(LocationDAO.findLocationByName(anyString())).thenReturn(new Location());
+        PowerMockito.when(LocationDAO.findLocationByName(anyString())).thenReturn(observable);
+
```
Comment (AdamGrzybkowski) [inline openmrs-client/src/test/java/org/openmrs/mobile/test/presenters/PatientDashboardVisitsPresenterTest.java]: I'm actually curious why LocationDAO methods are static... any idea  ?

File: openmrs-client/src/main/java/org/openmrs/mobile/dao/LocationDAO.java [hunk 5]
```diff
@@ -39,46 +43,50 @@ public void deleteAllLocations() {
...
     }
 
-    public static Location findLocationByName(String name) {
-        if(!StringUtils.notNull(name)){
-            return null;
-        }
-        Location location = new Location();
-        String where = String.format("%s = ?", LocationTable.Column.DISPLAY);
-        String[] whereArgs = new String[]{name};
+    public static Observable<Location> findLocationByName(String name) {
```
Comment (AdamGrzybkowski) [inline openmrs-client/src/main/java/org/openmrs/mobile/dao/LocationDAO.java]: let's revert this :)

Comment (AdamGrzybkowski) [inline openmrs-client/src/main/java/org/openmrs/mobile/dao/LocationDAO.java]: let's revert this :)

File: openmrs-client/src/test/java/org/openmrs/mobile/test/presenters/PatientDashboardVisitsPresenterTest.java (hunk 4 above)
Comment (shivtej1505) [inline openmrs-client/src/test/java/org/openmrs/mobile/test/presenters/PatientDashboardVisitsPresenterTest.java]: I don't think its necessary, I will remove that.

**Comment by shivtej1505** [comment]
@AdamGrzybkowski , PR updated!

**Pull request merged**